### ISSUE TITLE
[wasm][aot] StringSplitTests and ArrayTests failing with NullReferenceException

### ISSUE THREAD

**Issue by carlossanlop** [dotnet/runtime#105362]
## Build Information

Build: https://dev.azure.com/dnceng-public/public/_build/results?buildId=752664
Build error leg or test failing: `System.Tests.StringSplitTests.SplitCharSeparator` and `System.Tests.ArrayTests.Sort_Array_Array_NonGeneric`
[log_with_stack.txt](https://github.com/user-attachments/files/16428809/different_stack.txt)

## Error Message

```json
{
  "ErrorMessage" : ["Exception thrown during comparison", "Object reference not set to an instance of an object", "System.Tests"],
  "BuildRetry" : false,
  "ExcludeConsoleLog" : false
}
```

- PR: https://github.com/dotnet/runtime/pull/105354
- Queue: `browser-wasm linux Release LibraryTests_Smoke_AOT`
- Job result: https://dev.azure.com/dnceng-public/public/_build/results?buildId=752664&view=logs&j=62b70ecb-649d-5f05-1138-74c7e6e4705f&t=30431c62-b45b-56b6-e123-ff9f1cdc1425&l=59
- Log file: https://helixre107v0xd1eu3ibi6ka.blob.core.windows.net/dotnet-runtime-refs-pull-105354-merge-c9a9365ba7a34cf4b2/WasmTestOnV8-ST-System.Runtime.Tests/1/console.bfd09aa0.log?helixlogtype=result
- Output:
```
[22:22:37] info: [FAIL] System.Tests.StringSplitTests.SplitCharSeparator(value: "first,second", separator: ',', count: 4, options: None, expected: ["first", "second"])
[22:22:37] info: Assert.Equal() Failure: Exception thrown during comparison
[22:22:37] info:            ↓ (pos 0)
[22:22:37] info: Expected: [null, "second"]
[22:22:37] info: Actual:   ["first", "second"]
[22:22:37] info:            ↑ (pos 0)
[22:22:37] info: ---- System.NullReferenceException : Object reference not set to an instance of an object.
[22:22:37] info:    at System.Object.InvokeStub_StringSplitTests.SplitCharSeparator(Object , Span`1 )
[22:22:37] info:    at System.Reflection.MethodBaseInvoker.InvokeWithManyArgs(Object , BindingFlags , Binder , Object[] , CultureInfo )
[22:22:37] info: ----- Inner Stack Trace -----
[22:22:37] info: 
```

```
[22:22:43] info: [FAIL] System.Tests.ArrayTests.Sort_Array_Array_NonGeneric(keys: [3, 5, 6, 6], items: [4, 7, 9, 3], index: 0, length: 4, comparer: null, expectedKeys: [3, 5, 6, 6], expectedItems: [4, 7, 9, 3])
[22:22:43] info: Assert.Equal() Failure: Exception thrown during comparison
[22:22:43] info:            ↓ (pos 0)
[22:22:43] info: Expected: [null, 7, 9, 3]
[22:22:43] info: Actual:   [4, 7, 9, 3]
[22:22:43] info:            ↑ (pos 0)
[22:22:43] info: ---- System.NullReferenceException : Object reference not set to an instance of an object.
[22:22:43] info:    at System.Object.InvokeStub_ArrayTests.Sort_Array_Array_NonGeneric(Object , Span`1 )
[22:22:43] info:    at System.Reflection.MethodBaseInvoker.InvokeWithManyArgs(Object , BindingFlags , Binder , Object[] , CultureInfo )
[22:22:43] info: ----- Inner Stack Trace -----
[22:22:43] info: 
```

 ### Known issue validation
**Build: :mag_right:** https://dev.azure.com/dnceng-public/public/_build/results?buildId=752664
**Error message validated:** `[Exception thrown during comparison Object reference not set to an instance of an object System.Tests`]
**Result validation:** :white_check_mark: Known issue matched with the provided build.
**Validation performed at:** 8/2/2024 7:39:42 AM UTC

### Report

|Build|Definition|Test|Pull Request|
|---|---|---|---|
|[773753](https://dev.azure.com/dnceng-public/public/_build/results?buildId=773753)|dotnet/runtime|[WasmTestOnV8-ST-System.Runtime.Tests.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=773753&view=ms.vss-test-web.build-test-results-tab&runId=19783786&resultId=109440)|dotnet/runtime#106238|
|[773642](https://dev.azure.com/dnceng-public/public/_build/results?buildId=773642)|dotnet/runtime|[WasmTestOnV8-ST-System.Runtime.Tests.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=773642&view=ms.vss-test-web.build-test-results-tab&runId=19779974&resultId=109440)|dotnet/runtime#106303|
|[773607](https://dev.azure.com/dnceng-public/public/_build/results?buildId=773607)|dotnet/runtime|[WasmTestOnV8-ST-System.Runtime.Tests.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=773607&view=ms.vss-test-web.build-test-results-tab&runId=19778906&resultId=109440)|dotnet/runtime#105597|
|[772782](https://dev.azure.com/dnceng-public/public/_build/results?buildId=772782)|dotnet/runtime|[WasmTestOnV8-ST-System.Runtime.Tests.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=772782&view=ms.vss-test-web.build-test-results-tab&runId=19755220&resultId=109440)|dotnet/runtime#106267|
|[772725](https://dev.azure.com/dnceng-public/public/_build/results?buildId=772725)|dotnet/runtime|[WasmTestOnV8-ST-System.Runtime.Tests.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=772725&view=ms.vss-test-web.build-test-results-tab&runId=19753884&resultId=109440)|dotnet/runtime#106261|
|[771669](https://dev.azure.com/dnceng-public/public/_build/results?buildId=771669)|dotnet/runtime|[WasmTestOnV8-ST-System.Runtime.Tests.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=771669&view=ms.vss-test-web.build-test-results-tab&runId=19715332&resultId=109440)|dotnet/runtime#106215|
|[771070](https://dev.azure.com/dnceng-public/public/_build/results?buildId=771070)|dotnet/runtime|[WasmTestOnV8-ST-System.Runtime.Tests.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=771070&view=ms.vss-test-web.build-test-results-tab&runId=19698694&resultId=109440)|dotnet/runtime#106192|
|[770672](https://dev.azure.com/dnceng-public/public/_build/results?buildId=770672)|dotnet/runtime|[WasmTestOnV8-ST-System.Runtime.Tests.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=770672&view=ms.vss-test-web.build-test-results-tab&runId=19686028&resultId=109440)|dotnet/runtime#106178|
|[770443](https://dev.azure.com/dnceng-public/public/_build/results?buildId=770443)|dotnet/runtime|[WasmTestOnV8-ST-System.Runtime.Tests.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=770443&view=ms.vss-test-web.build-test-results-tab&runId=19679300&resultId=100001)|dotnet/runtime#106167|
|[770096](https://dev.azure.com/dnceng-public/public/_build/results?buildId=770096)|dotnet/runtime|[WasmTestOnV8-ST-System.Runtime.Tests.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=770096&view=ms.vss-test-web.build-test-results-tab&runId=19668180&resultId=109440)|dotnet/runtime#105941|
|[769484](https://dev.azure.com/dnceng-public/public/_build/results?buildId=769484)|dotnet/runtime|[WasmTestOnV8-ST-System.Runtime.Tests.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=769484&view=ms.vss-test-web.build-test-results-tab&runId=19647786&resultId=109440)|dotnet/runtime#106116|
|[768855](https://dev.azure.com/dnceng-public/public/_build/results?buildId=768855)|dotnet/runtime|[WasmTestOnV8-ST-System.Runtime.Tests.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=768855&view=ms.vss-test-web.build-test-results-tab&runId=19622802&resultId=109440)|dotnet/runtime#105661|
|[768701](https://dev.azure.com/dnceng-public/public/_build/results?buildId=768701)|dotnet/runtime|[WasmTestOnV8-ST-System.Runtime.Tests.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=768701&view=ms.vss-test-web.build-test-results-tab&runId=19617476&resultId=109440)|dotnet/runtime#105713|
|[767716](https://dev.azure.com/dnceng-public/public/_build/results?buildId=767716)|dotnet/runtime|[WasmTestOnV8-ST-System.Runtime.Tests.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=767716&view=ms.vss-test-web.build-test-results-tab&runId=19579606&resultId=109439)|dotnet/runtime#106040|
|[767065](https://dev.azure.com/dnceng-public/public/_build/results?buildId=767065)|dotnet/runtime|[WasmTestOnV8-ST-System.Runtime.Tests.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=767065&view=ms.vss-test-web.build-test-results-tab&runId=19553104&resultId=109439)|dotnet/runtime#105867|
|[767024](https://dev.azure.com/dnceng-public/public/_build/results?buildId=767024)|dotnet/runtime|[WasmTestOnV8-ST-System.Runtime.Tests.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=767024&view=ms.vss-test-web.build-test-results-tab&runId=19551252&resultId=109439)|dotnet/runtime#105946|
|[765792](https://dev.azure.com/dnceng-public/public/_build/results?buildId=765792)|dotnet/runtime|[WasmTestOnV8-ST-System.Runtime.Tests.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=765792&view=ms.vss-test-web.build-test-results-tab&runId=19511024&resultId=109439)|dotnet/runtime#99596|
|[764956](https://dev.azure.com/dnceng-public/public/_build/results?buildId=764956)|dotnet/runtime|[WasmTestOnV8-ST-System.Runtime.Tests.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=764956&view=ms.vss-test-web.build-test-results-tab&runId=19465522&resultId=109439)|dotnet/runtime#105909|
|[764868](https://dev.azure.com/dnceng-public/public/_build/results?buildId=764868)|dotnet/runtime|[WasmTestOnV8-ST-System.Runtime.Tests.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=764868&view=ms.vss-test-web.build-test-results-tab&runId=19459576&resultId=144071)||
|[764197](https://dev.azure.com/dnceng-public/public/_build/results?buildId=764197)|dotnet/runtime|[WasmTestOnV8-ST-System.Runtime.Tests.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=764197&view=ms.vss-test-web.build-test-results-tab&runId=19434938&resultId=109439)|dotnet/runtime#105876|
|[761037](https://dev.azure.com/dnceng-public/public/_build/results?buildId=761037)|dotnet/runtime|[WasmTestOnV8-ST-System.Runtime.Tests.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=761037&view=ms.vss-test-web.build-test-results-tab&runId=19327284&resultId=109439)|dotnet/runtime#105671|
|[760113](https://dev.azure.com/dnceng-public/public/_build/results?buildId=760113)|dotnet/runtime|[WasmTestOnV8-ST-System.Runtime.Tests.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=760113&view=ms.vss-test-web.build-test-results-tab&runId=19309206&resultId=109439)|dotnet/runtime#105561|
|[759296](https://dev.azure.com/dnceng-public/public/_build/results?buildId=759296)|dotnet/runtime|[WasmTestOnV8-ST-System.Runtime.Tests.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=759296&view=ms.vss-test-web.build-test-results-tab&runId=19278478&resultId=109439)|dotnet/runtime#105525|
|[759359](https://dev.azure.com/dnceng-public/public/_build/results?buildId=759359)|dotnet/runtime|[WasmTestOnV8-ST-System.Runtime.Tests.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=759359&view=ms.vss-test-web.build-test-results-tab&runId=19275756&resultId=109439)|dotnet/runtime#105637|
|[759349](https://dev.azure.com/dnceng-public/public/_build/results?buildId=759349)|dotnet/runtime|[WasmTestOnV8-ST-System.Runtime.Tests.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=759349&view=ms.vss-test-web.build-test-results-tab&runId=19275632&resultId=100001)|dotnet/runtime#105636|
|[758744](https://dev.azure.com/dnceng-public/public/_build/results?buildId=758744)|dotnet/runtime|[WasmTestOnV8-ST-System.Runtime.Tests.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=758744&view=ms.vss-test-web.build-test-results-tab&runId=19257252&resultId=109439)|dotnet/runtime#105559|
|[758343](https://dev.azure.com/dnceng-public/public/_build/results?buildId=758343)|dotnet/runtime|[WasmTestOnV8-ST-System.Runtime.Tests.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=758343&view=ms.vss-test-web.build-test-results-tab&runId=19243398&resultId=141242)||
|[757794](https://dev.azure.com/dnceng-public/public/_build/results?buildId=757794)|dotnet/runtime|[WasmTestOnV8-ST-System.Runtime.Tests.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=757794&view=ms.vss-test-web.build-test-results-tab&runId=19224422&resultId=109439)|dotnet/runtime#105591|
|[756377](https://dev.azure.com/dnceng-public/public/_build/results?buildId=756377)|dotnet/runtime|[WasmTestOnV8-ST-System.Runtime.Tests.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=756377&view=ms.vss-test-web.build-test-results-tab&runId=19176810&resultId=109439)|dotnet/runtime#105543|
|[756323](https://dev.azure.com/dnceng-public/public/_build/results?buildId=756323)|dotnet/runtime|[WasmTestOnV8-ST-System.Runtime.Tests.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=756323&view=ms.vss-test-web.build-test-results-tab&runId=19174272&resultId=109439)|dotnet/runtime#105540|
|[755983](https://dev.azure.com/dnceng-public/public/_build/results?buildId=755983)|dotnet/runtime|[WasmTestOnV8-ST-System.Runtime.Tests.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=755983&view=ms.vss-test-web.build-test-results-tab&runId=19166752&resultId=109439)|dotnet/runtime#105524|
|[755382](https://dev.azure.com/dnceng-public/public/_build/results?buildId=755382)|dotnet/runtime|[WasmTestOnV8-ST-System.Runtime.Tests.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=755382&view=ms.vss-test-web.build-test-results-tab&runId=19151030&resultId=100001)|dotnet/runtime#105432|
|[754933](https://dev.azure.com/dnceng-public/public/_build/results?buildId=754933)|dotnet/runtime|[WasmTestOnV8-ST-System.Runtime.Tests.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=754933&view=ms.vss-test-web.build-test-results-tab&runId=19137982&resultId=109439)|dotnet/runtime#105469|
|[752834](https://dev.azure.com/dnceng-public/public/_build/results?buildId=752834)|dotnet/runtime|[WasmTestOnV8-ST-System.Runtime.Tests.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=752834&view=ms.vss-test-web.build-test-results-tab&runId=19073038&resultId=109439)|dotnet/runtime#105351|
|[752664](https://dev.azure.com/dnceng-public/public/_build/results?buildId=752664)|dotnet/runtime|[WasmTestOnV8-ST-System.Runtime.Tests.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=752664&view=ms.vss-test-web.build-test-results-tab&runId=19069520&resultId=109439)|dotnet/runtime#105354|
#### Summary
|24-Hour Hit Count|7-Day Hit Count|1-Month Count|
|---|---|---|
|0|0|35|

**Comment by maraf** [comment]
@ilonatommy Can you please try to look what is going on? There isn't any change in the test suite or String.Mono.cs recently.

**Comment by ilonatommy** [comment]
```
Using js engine V8 from path /home/helixbot/.jsvu/bin/v8
12.8.364
```
[log_array.txt](https://github.com/user-attachments/files/16411412/log_array.txt)

**Comment by ilonatommy** [comment]
Might be connected: https://github.com/dotnet/runtime/issues/101481 (same stack starting from `assert_Xunit_Assert_Equal_T_REF_T_REF_T_REF`)

**Comment by ilonatommy** [comment]
I am not 100% sure if the log is connected with the null-valued first element issue, the runtime where I got it is debug-based and it's more prone to throwing. What config are you using? Does it reproduce every time? For me, I got that stack that I pasted only once, then it always passed both in debug and release

**Comment by mkhamoyan** [comment]
> I am not 100% sure if the log is connected with the null-valued first element issue, the runtime where I got it is debug-based and it's more prone to throwing. What config are you using? Does it reproduce every time? For me, I got that stack that I pasted only once, then it always passed both in debug and release

It is with release, happened first time.

**Comment by pavelsavara** [comment]
Last few [logs](https://helixre107v0xdcypoyl9e7f.blob.core.windows.net/dotnet-runtime-refs-pull-105867-merge-d2f88a24f8e9439c97/WasmTestOnV8-ST-System.Runtime.Tests/1/console.d4c2f57f.log?helixlogtype=result) have `Unsupported long branch opcode` in them which is https://github.com/dotnet/runtime/issues/106028
```
[21:08:34] info: Discovering: System.Runtime.Tests.dll (method display = ClassAndMethod, method display options = None)
[21:08:35] info: MONO_WASM: GetGrouping:0 code generation failed: Error: Unsupported long branch opcode: bne.un.i4 Error: Unsupported long branch opcode: bne.un.i4
[21:08:35] info:     at di (/root/helix/work/workitem/e/wasm_build/AppBundle/_framework/dotnet.runtime.js:3:112931)
[21:08:35] info:     at fi (/root/helix/work/workitem/e/wasm_build/AppBundle/_framework/dotnet.runtime.js:3:114018)
[21:08:35] info:     at /root/helix/work/workitem/e/wasm_build/AppBundle/_framework/dotnet.runtime.js:3:163954
[21:08:35] info:     at Object.generator (/root/helix/work/workitem/e/wasm_build/AppBundle/_framework/dotnet.runtime.js:3:164741)
[21:08:35] info:     at Ns.emitImportsAndFunctions (/root/helix/work/workitem/e/wasm_build/AppBundle/_framework/dotnet.runtime.js:3:73961)
[21:08:35] info:     at /root/helix/work/workitem/e/wasm_build/AppBundle/_framework/dotnet.runtime.js:3:164807
[21:08:35] info:     at Cc (/root/helix/work/workitem/e/wasm_build/AppBundle/_framework/dotnet.runtime.js:3:165877)
[21:08:35] info:     at dotnet.native.wasm.mono_interp_tier_prepare_jiterpreter_fast (wasm://wasm/dotnet.native.wasm-103598ee:wasm-function[116641]:0x1c1e312)
```

**Comment by pavelsavara** [comment]
I hope the the root cause is https://github.com/dotnet/runtime/issues/106200 and fix https://github.com/dotnet/runtime/pull/106313

**Comment by lewing** [comment]
> I hope the the root cause is #106200 and fix #106313

🤞 looking good right now
